---
title: "HW1"
author: "Huiyu Hu"
date: "1/18/2018"
output:
  html_document: default
  pdf_document: default
---


# Part 1 
 * Finished the GitHub setting.
 
# Part 2

 1. How many persons are in the data set (statisticians call this n)? How many SNPs are in the data set (statisticians call this p)?
 

```{bash}
awk 'END { print NR }' /home/m280-data/hw1/merge-geno.fam
```
```{bash}
awk 'END { print NR }' /home/m280-data/hw1/merge-geno.bim
```
 * Answer: n = 959; p = 8348674

 2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?
```{bash}
 awk < /home/m280-data/hw1/merge-geno.bim '{print $1}' | uniq -c
```
  + Answer:  

 3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp â€“ 48,130,769 bp. How many SNPs are located within MAP4 gene?
```{bash}
awk '{if ($1 == 3 && $4 >= 47892180 && $4 <= 48130769) print}' /home/m280-data/hw1/merge-geno.bim | wc -l
```
 * Answer: There are 894 SNPs lovated within MAP4 gene.
 4. Reformat
 * For .bim file:
```{bash}
 echo "    2.40  = FILE FORMAT VERSION NUMBER." > /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.lalala
 echo '8348674  = NUMBER OF SNPS LISTED HERE.' >> /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.lalala
 awk '{OFS = ","} {print $2, $1, $4}' /home/m280-data/hw1/merge-geno.bim  >> /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.lalala
 head /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.lalala
```
 * For .fam file
 * Comments: 1, change seperator to ","; 2, strip the string T2DG from the IDs; 3, convert sex from number to char; 4, remove missing value.
          
```{bash}
awk '{OFS = ","} {print $1, $2, $3, $4, $5, $6}' /home/m280-data/hw1/merge-geno.fam | sed 's/T2DG//g' | awk -F, -v sex="" '{OFS = ","} {if ($5 == 1) sex = "M"; else sex = "F"} {print $1,$2,$3,$4, sex,$6}' | awk -F, '{OFS=","} { for (i = 1; i <= NF; i++) {$i = ($i == 0 ? "" : $i)}; print }'  > /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.hehehe 
head -20 /home/huiyuhu/biostat-m280-2018-winter/hw1/merge-geno.hehehe
```
 
# Part 3 
1. 
```{r}
  ## parsing command arguments
  for (arg in commandArgs(TRUE)) {
    eval(parse(text=arg))
  }
  
  ## check if a given integer is prime
  isPrime = function(n) {
    if (n <= 3) {
      return (TRUE)
    }
    if (any((n %% 2:floor(sqrt(n))) == 0)) {
      return (FALSE)
    }
    return (TRUE)
  }
  
  ## estimate mean only using observation with prime indices
  estMeanPrimes = function (x) {
    n = length(x)
    ind = sapply(1:n, isPrime)
    return (mean(x[ind]))
  }
  
  
runSim <- function(seed, n, dist, rep){
    # step 1: set random seed
    set.seed(seed)
    
  sum1 <- 0
  sum2 <- 0
  for ( i in 1:rep) {
    # step 2: generate data according to argument dist
    if (dist == "gaussian"){
      x = rnorm(n,0,1)
    }
    else if (dist == "t1"){
      x = rt(n, df=1)
    }
    else if (dist == "t5"){
      x = rt(n, df=5)
    }
    
    # estimate mean
    emp <- estMeanPrimes(x)
    emc <- mean(x)
    sum1 <- sum1 + (emp-0)^2
    sum2 <- sum2 + (emc-0)^2
  }
  
  result <- paste("MSE (primed-indexed average estimator) =",sum1/rep,  ";  MSE (classical sample average estimator) =", sum2/rep, sep = " ")
  return(result)

}

#test
runSim(200,5000,"gaussian",50)
```
 
 
