---
title: "Biostat M280 Homework 2"
subtitle: Due Feb 16 @ 11:59PM
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Q1

Read [Chapter 7](http://r4ds.had.co.nz/exploratory-data-analysis.html) (Exploratory Data Analysis) of _R for Data Science_ and do exercises 7.3.4, 7.4.1, 7.5.1.1, 7.5.2.1, and 7.5.3.1.
#### exercises 7.3.4
1.Explore the distribution of each of the x, y, and z variables in diamonds. What do you learn? Think about a diamond and how you might decide which dimension is the length, width, and depth.
```{r}
library("tidyverse")
ggplot(data = diamonds) +
  geom_bar(mapping = aes(x = x))
ggplot(data = diamonds) +
  geom_bar(mapping = aes(x = y))
ggplot(data = diamonds) +
  geom_bar(mapping = aes(x = z))
```

2. Explore the distribution of price. Do you discover anything unusual or surprising? (Hint: Carefully think about the binwidth and make sure you try a wide range of values.)

```{r}
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = price), binwidth = 40)
```

3. How many diamonds are 0.99 carat? How many are 1 carat? What do you think is the cause of the difference?
```{r}
filter(diamonds, carat == 0.99) %>%
  summarise(n())
```
```{r}
filter(diamonds, carat == 1) %>%
  summarise(n())
```
4. Compare and contrast coord_cartesian() vs xlim() or ylim() when zooming in on a histogram. What happens if you leave binwidth unset? What happens if you try and zoom so only half a bar shows?

```{r}
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = y),binwidth = 0.5) +
  coord_cartesian(ylim = c(0, 100))
# What happens if you leave binwidth unset?
ggplot(diamonds) + 
  geom_histogram(mapping = aes(x = y)) +
  coord_cartesian(ylim = c(0, 100))
# What happens if you try and zoom so only half a bar shows?

```

#### exercises 7.4.1
1. What happens to missing values in a histogram? What happens to missing values in a bar chart? Why is there a difference?
```{r}
diamonds %>%
  ggplot(aes(price)) +
  geom_histogram(bins = 500)

diamonds %>%
  ggplot(aes(price)) +
  geom_bar()
```
* Answer: 

2. What does na.rm = TRUE do in mean() and sum()?
* Answer: 

#### 7.5.1.1
1. Use what you’ve learned to improve the visualisation of the departure times of cancelled vs. non-cancelled flights.
```{r}
library("nycflights13")
nycflights13::flights %>% 
  mutate(
    cancelled = is.na(dep_time),
    sched_hour = sched_dep_time %/% 100,
    sched_min = sched_dep_time %% 100,
    sched_dep_time = sched_hour + sched_min / 60
  ) %>% 
  ggplot(mapping = aes(sched_dep_time)) + 
    geom_freqpoly(mapping = aes(colour = cancelled), binwidth = 1/4)
```



## Q2 (optional)

Read [Chapter 23](http://r4ds.had.co.nz/model-basics.html) (Model Basics) and [Chapter 24](http://r4ds.had.co.nz/model-building.html) (Model Building) of _R for Data Science_ and do exercises 24.2.3 and 24.3.5.

## Q3

Redo HW1 Q2 using tidyverse.
* Read in data
```{r}
library("tidyverse")
geno_data1 <- read_tsv("/home/m280-data/hw1/merge-geno.bim", col_names = F) 
geno_data1 <- as_data_frame(geno_data1)
colnames(geno_data1) <- c("Chromosome","SNP_ID","Genetic_Distance","bp",
                       "Allele 1","Allele 2")
geno_data2 <- read_delim("/home/m280-data/hw1/merge-geno.fam"," ", col_names = F)
geno_data2 <- as_data_frame(geno_data2)
colnames(geno_data2) <- c("Family","Person","Father","Mother",
                       "Sex","Affection")
```
1. How many persons are in the data set (statisticians call this n)? How many SNPs are in the data set (statisticians call this p)?

```{r}
geno_data1 %>% 
  count()
geno_data2 %>% 
  count()
```
* Answer: n = $959$ ; p = $8348674$. 

2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?

```{r}
geno_data1 %>% 
  group_by(Chromosome) %>% 
  summarise(n = n()) 
```
* Answer:

| $Chromesome$ | $SNPs$ |
|--------------|--------|
|     1        |1309299 |
|     3        |1215399 |
|     5        |1090185 |
|     7        | 980944 |
|     9        | 732013 |
|     11       | 815860 |
|     13       | 602809 |
|     15       | 491208 |
|     17       | 477990 |
|     19       | 393615 |
|     21       | 239352 |


3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp – 48,130,769 bp. How many SNPs are located within MAP4 gene?
```{r}
filter(geno_data1, Chromosome == 3, bp >= 47892180, bp <= 48130769 ) %>% 
  summarise(n = n()) 
```
* Answer: There are 894 SNPs located within MAP4 gene.

4. Reformat:
* Part 1:
```{r}
output_file <- "/home/huiyuhu/biostat-m280-2018-winter/hw2/mendel1.txt"
fileConn <- file(output_file, open = "a")
writeLines("    2.40 = FILE FORMAT VERSION NUMBER", fileConn)
writeLines("8348674 = NUMBER OF SNPS LISTED HERE", fileConn)
close(fileConn)

mendel <- geno_data1[, c("SNP_ID", "Chromosome","bp")]
write_delim(mendel, output_file, delim = ",", col_names = FALSE, append = TRUE)
```
```{bash}
head /home/huiyuhu/biostat-m280-2018-winter/hw2/mendel1.txt
```

* Part 2:
```{r}
geno_data2 %>%
  mutate(Sex = replace(Sex, Sex == 1, "M")) %>%
  mutate(Sex = replace(Sex, Sex == 2, "F")) %>%
  mutate(Father = replace(Father, Father == "0", NA)) %>%
  mutate(Mother = replace(Mother, Mother == "0", NA)) %>%
  mutate(Affection = replace(Affection, Affection == 0, NA))
```


## Q4 (optional)

Redo HW1 Q3 on Hoffman2, except now we want to submit each `runSum.R` job to a different node in the cluster.
